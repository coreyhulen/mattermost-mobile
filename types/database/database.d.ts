// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import { AppSchema, Database } from '@nozbe/watermelondb';
import Model from '@nozbe/watermelondb/Model';
import { Migration } from '@nozbe/watermelondb/Schema/migrations';
import { Class } from '@nozbe/watermelondb/utils/common';

import { IsolatedEntities, OperationType } from '../../app/database/admin/data_operator';
import { DatabaseType } from '../../app/database/admin/database_manager';

export type MigrationEvents = {
    onSuccess: () => void;
    onStarted: () => void;
    onFailure: (error: string) => void;
};

export type MMAdaptorOptions = {
    dbPath: string;
    schema: AppSchema;
    migrationSteps?: Migration[];
    migrationEvents?: MigrationEvents;
};

export type MMDatabaseConnection = {
    actionsEnabled?: boolean;
    dbName: string;
    dbType?: DatabaseType.DEFAULT | DatabaseType.SERVER;
    serverUrl?: string;
};

export type DefaultNewServer = {
    databaseFilePath: string;
    displayName: string;
    serverUrl: string;
};

// A database connection is of type 'Database'; unless it fails to be initialize and in which case it becomes 'undefined'
export type DatabaseInstance = Database | undefined;

export type RawApp = {
    buildNumber: string;
    createdAt: number;
    id: string;
    versionNumber: string;
};

export type RawGlobal = {
    id: string;
    name: string;
    value: string;
};

export type RawServers = {
    dbPath: string;
    displayName: string;
    id: string;
    mentionCount: number;
    unreadCount: number;
    url: string;
};

export type RawCustomEmoji = {
    id: string;
    name: string;
    create_at: number;
    update_at: number;
    delete_at: number;
    creator_id: string;
};

export type RawRole = {
    id: string;
    name: string;
    permissions: [];
};

export type RawSystem = {
    id: string;
    name: string;
    value: string;
};

export type RawTermsOfService = {
    id: string;
    acceptedAt: number;
};

export type RawEmbeds = [{ type: string; url: string; data: {} }];

export type RawEmojis = [
    { id: string; creator_id: string; name: string; create_at: number; update_at: number; delete_at: number },
];

export type RawFile = {
    create_at: number;
    delete_at: number;
    extension: string;
    has_preview_image?: boolean;
    height: number;
    id?: string;
    localPath?: string;
    mime_type?: string;
    mini_preview?: string; // thumbnail
    name: string;
    post_id: string;
    size: number;
    update_at: number;
    user_id: string;
    width?: number;
};

export type RawReaction = {
    id?: string;
    create_at: number;
    delete_at: number;
    emoji_name: string;
    post_id: string;
    update_at: number;
    user_id: string;
};

// interface PostMetadataTypes {
//     embeds: Array<PostEmbed>;
//     emojis: Array<CustomEmoji>;
//     files: Array<FileInfo>;
//     images: Dictionary<PostImage>;
//     reactions: Array<Reaction>;
// }

// The RawPost describes the shape of the object received from a getPosts request
export type RawPost = {
    channel_id: string;
    create_at: number;
    delete_at: number;
    edit_at: number;
    file_ids: string[];
    filenames: string[];
    hashtags: string;
    id: string;
    is_pinned?: boolean;
    last_reply_at?: number;
    message: string;
    original_id: string;
    parent_id: string;
    participants?: null;
    pending_post_id: string;
    prev_post_id?: string; // taken from getPosts API call; outside of post object
    props: object;
    reply_count?: 0;
    root_id: string;
    type: string;
    update_at: number;
    user_id: string;
    metadata: {
        embeds: RawEmbeds;
        emojis: RawEmojis;
        files: RawFile[];
        images: {};
        reactions: RawReaction[];
    };
};

export type RawChannelMembers = {
    channel_id: string;
    explicit_roles: string;
    last_update_at: number;
    last_viewed_at: number;
    mention_count: number;
    msg_count: number;
    notify_props: NotifyProps;
    roles: string;
    scheme_admin: boolean;
    scheme_guest: boolean;
    scheme_user: boolean;
    user_id: string;
};

export type ChannelType = 'D' | 'O' | 'G' | 'P';

export type RawChannel = {
    create_at: number;
    creator_id: string;
    delete_at: number;
    display_name: string;
    extra_update_at: number;
    group_constrained: boolean | null;
    header: string;
    id: string;
    last_post_at: number;
    name: string;
    props: null;
    purpose: string;
    scheme_id: null;
    shared: null;
    team_id: string;
    total_msg_count: number;
    type: ChannelType;
    update_at: number;
};

export type RawPostsInThread = {
    id?: string;
    earliest: number;
    latest?: number;
    post_id: string;
};

export type RecordValue =
    | RawApp
    | RawCustomEmoji
    | RawFile
    | RawGlobal
    | RawPost
    | RawPostsInThread
    | RawReaction
    | RawRole
    | RawServers
    | RawSystem
    | RawTermsOfService;

export type DataFactory = {
    database: Database;
    generator?: (model: Model) => void;
    optType?: OperationType;
    tableName?: string;
    value: RecordValue;
};

export type HandleBaseData = {
    database?: Database;
    optType: OperationType;
    tableName: string;
    values: RecordValue[];
    recordOperator: (recordOperator: {
        optType: OperationType;
        value: RecordValue;
        database: Database;
        tableName: string;
    }) => void;
};

export type BatchOperations = { database: Database; models: Model[] };

export type HandleIsolatedEntityData = { optType: OperationType; tableName: IsolatedEntities; values: RecordValue[] };

export type Models = Class<Model>[];

// The elements needed to create a new connection
export type DatabaseConnection = {
    databaseConnection: MMDatabaseConnection;
    shouldAddToDefaultDatabase: boolean;
};

// The elements required to switch to another active server database
export type ActiveServerDatabase = { displayName: string; serverUrl: string };
