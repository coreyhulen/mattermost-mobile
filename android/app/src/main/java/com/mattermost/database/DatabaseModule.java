package com.mattermost.database;

import android.database.Cursor;
import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.nozbe.watermelondb.Database;


public class DatabaseModule extends ReactContextBaseJavaModule {
    private static Database defaultDatabase;
    private ReactApplicationContext reactContext;

    public DatabaseModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "MattermostDatabase";
    }

    @ReactMethod
    public void setDefaultDatabase(String databaseName) {
        this.defaultDatabase = new Database(databaseName, reactContext);
    }

    public static String getOnlyServerUrl() {
        if (defaultDatabase != null) {
            String emptyArray[] = {};
            String query = "SELECT url FROM servers";
            Cursor cursor = defaultDatabase.rawQuery(query, emptyArray);

            if (cursor.getCount() == 1) {
                cursor.moveToFirst();
                return cursor.getString(0);
            }
        }

        return null;
    }
}
