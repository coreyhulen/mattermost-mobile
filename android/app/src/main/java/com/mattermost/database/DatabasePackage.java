package com.mattermost.database;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;
import com.mattermost.rnbeta.MainApplication;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class DatabasePackage implements ReactPackage {
    MainApplication mApplication;

    public DatabasePackage(MainApplication application) {
        mApplication = application;
    }

    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
        return Arrays.<NativeModule>asList(new DatabaseModule(reactContext));
    }

    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }
}
