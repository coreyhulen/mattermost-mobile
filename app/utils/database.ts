// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Client4} from '@client/rest';
import DatabaseConnectionException from '@database/exceptions/database_connection_exception';
import DatabaseManager from '@database/manager';
import urlParse from 'url-parse';

type SetActiveDatabaseArgs = {
  serverUrl: string;
  displayName?: string;
};

export const createAndSetActiveDatabase = async ({serverUrl, displayName}: SetActiveDatabaseArgs) => {
    const connectionName = displayName ?? urlParse(Client4.getUrl())?.hostname;

    try {
        const databaseClient = new DatabaseManager();
        await databaseClient.getDatabaseConnection({serverUrl, connectionName, setAsActiveDatabase: true});
    } catch (e) {
        throw new DatabaseConnectionException(
            `createAndSetActiveDatabase: Unable to create and set serverUrl ${serverUrl} as current active database with name ${displayName}`,
        );
    }
};

export const getDefaultDatabase = async () => {
    try {
        const databaseClient = new DatabaseManager();
        const defaultDatabase = await databaseClient.getDefaultDatabase();
        return {
            error: defaultDatabase ? null : 'Unable to retrieve the App database.',
            defaultDatabase,
        };
    } catch (e) {
        return {
            error: 'Unable to retrieve the App database.',
            defaultDatabase: null,
        };
    }
};

export const getActiveServerDatabase = async () => {
    try {
        const databaseClient = new DatabaseManager();
        const activeServerDatabase = await databaseClient.getActiveServerDatabase();

        return {
            error: activeServerDatabase ? null : 'Unable to retrieve the current active server database.',
            activeServerDatabase,
        };
    } catch (e) {
        return {
            error: 'Unable to retrieve the current active server database.',
            activeServerDatabase: null,
        };
    }
};
