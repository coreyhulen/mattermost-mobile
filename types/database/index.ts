// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export interface NotifyProps {
    channel: true;
    desktop: string;
    desktop_sound: true;
    email: true;
    first_name: true
    mention_keys: string;
    push: string;
}

export interface UserProps {
    [userPropsName : string] : any
}

export interface Timezone {
    automaticTimezone: string
    manualTimezone: string,
    useAutomaticTimezone: true,
}
