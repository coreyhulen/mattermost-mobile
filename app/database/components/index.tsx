// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {ComponentType, useEffect, useState} from 'react';
import {Database} from '@nozbe/watermelondb';
import DatabaseProvider from '@nozbe/watermelondb/DatabaseProvider';

import {MM_TABLES} from '@constants/database';
import DatabaseManager from '@database/manager';
import Servers from '@database/models/default/servers';

const {SERVERS} = MM_TABLES.DEFAULT;

export function withServerDatabase<T>(Component: ComponentType<T>): ComponentType<T> {
    return function ServerDatabaseComponent(props) {
        const [database, setDatabase] = useState<Database|unknown>();

        // If we don't need to await the async functions this side effect is not needed
        useEffect(() => {
            const observer = async (servers: Servers[]) => {
                const server = servers.reduce((a, b) => (a.lastActiveAt > b.lastActiveAt ? a : b));

                //fixme: 1: DatabaseManager should now be called as a client

                // The server database should already exists at this point
                // there should not be a need to await
                const serverDatabase = await DatabaseManager.retrieveDatabaseInstances([server.url]);
                if (serverDatabase?.[0]?.dbInstance) {
                    setDatabase(serverDatabase[0].dbInstance);
                }
            };

            const init = async () => {
                // TODO: At this point the database should be already present
                // there should not be a need to await

                const db = await DatabaseManager.getDefaultDatabase();
                db?.collections.
                    get(SERVERS).
                    query().
                    observeWithColumns(['last_active_at']).
                    subscribe(observer);
            };

            init();
        });

        if (!database) {
            return null;
        }

        return (
            <DatabaseProvider database={(database as Database)}>
                <Component {...props}/>
            </DatabaseProvider>
        );
    };
}
