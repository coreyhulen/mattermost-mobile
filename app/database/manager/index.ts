// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Database, Q} from '@nozbe/watermelondb';
import SQLiteAdapter from '@nozbe/watermelondb/adapters/sqlite';
import logger from '@nozbe/watermelondb/utils/common/logger';
import {DeviceEventEmitter, Platform} from 'react-native';
import {FileSystem} from 'react-native-unimodules';

import {MIGRATION_EVENTS, MM_TABLES} from '@constants/database';
import DefaultMigration from '@database/migration/default';
import {App, Global, Servers} from '@database/models/default';
import {defaultSchema} from '@database/schema/default';
import ServerMigration from '@database/migration/server';
import {
    Channel,
    ChannelInfo,
    ChannelMembership,
    CustomEmoji,
    Draft,
    File,
    Group,
    GroupMembership,
    GroupsInChannel,
    GroupsInTeam,
    MyChannel,
    MyChannelSettings,
    MyTeam,
    Post,
    PostMetadata,
    PostsInChannel,
    PostsInThread,
    Preference,
    Reaction,
    Role,
    SlashCommand,
    System,
    Team,
    TeamChannelHistory,
    TeamMembership,
    TeamSearchHistory,
    TermsOfService,
    User,
} from '@database/models/server';
import {serverSchema} from '@database/schema/server';
import {
    ActiveServerDatabaseArgs,
    DatabaseConnectionArgs,
    DatabaseInstance,
    DatabaseInstances,
    DefaultNewServerArgs,
    GetDatabaseConnectionArgs,
    MigrationEvents,
    Models,
    RetrievedDatabase,
} from '@typings/database/database';
import {DatabaseType} from '@typings/database/enums';
import IServers from '@typings/database/servers';
import {deleteIOSDatabase, getIOSAppGroupDetails} from '@utils/mattermost_managed';
import urlParse from 'url-parse';

const {SERVERS} = MM_TABLES.DEFAULT;
const DEFAULT_DATABASE = 'default';

class DatabaseManager {
  private activeDatabase: DatabaseInstance;
  private defaultDatabase: DatabaseInstance;
  private readonly defaultModels: Models;
  private readonly iOSAppGroupDatabase: string | null;
  private readonly androidFilesDirectory: string | null;
  private readonly serverModels: Models;

  constructor() {
      this.defaultModels = [App, Global, Servers];
      this.serverModels = [
          Channel,
          ChannelInfo,
          ChannelMembership,
          CustomEmoji,
          Draft,
          File,
          Group,
          GroupMembership,
          GroupsInChannel,
          GroupsInTeam,
          MyChannel,
          MyChannelSettings,
          MyTeam,
          Post,
          PostMetadata,
          PostsInChannel,
          PostsInThread,
          Preference,
          Reaction,
          Role,
          SlashCommand,
          System,
          Team,
          TeamChannelHistory,
          TeamMembership,
          TeamSearchHistory,
          TermsOfService,
          User,
      ];

      this.iOSAppGroupDatabase = Platform.OS === 'ios' ? getIOSAppGroupDetails().appGroupDatabase : null;
      this.androidFilesDirectory = Platform.OS === 'android' ? FileSystem.documentDirectory : null;
  }

  /**
     * getDatabaseConnection: Given a server url (serverUrl) and a flag (setAsActiveDatabase), this method will attempt
     * to retrieve an existing database connection previously created for that url.  If not found, it will create a new connection and register it in the DEFAULT_DATABASE
     * @param {string} serverUrl
     * @param {boolean} setAsActiveDatabase
     * @returns {Promise<DatabaseInstance>}
     */
  getDatabaseConnection = async ({serverUrl, setAsActiveDatabase}: GetDatabaseConnectionArgs) => {
      // We potentially already have this server registered; so we'll try to retrieve it if it is present under DEFAULT_DATABASE/GLOBAL entity
      const existingServers = await this.retrieveDatabaseInstances([serverUrl]) as RetrievedDatabase[];

      // Since we only passed one serverUrl, we'll expect only one value in the array
      const serverDatabase = existingServers?.[0];

      let connection: DatabaseInstance;
      let databaseName: string;

      if (serverDatabase) {
          // This serverUrl has previously been registered on the app
          databaseName = serverDatabase.displayName;
          connection = serverDatabase.dbInstance;
      } else {
          // Or, it might be that the user has this server on the web-app but not mobile-app; so we'll need to create a new entry for this new serverUrl
          databaseName = urlParse(serverUrl).hostname;
          connection = await this.createDatabaseConnection({
              shouldAddToDefaultDatabase: true,
              configs: {
                  actionsEnabled: true,
                  dbName: databaseName,
                  dbType: DatabaseType.SERVER,
                  serverUrl,
              },
          });
      }

      if (setAsActiveDatabase) {
          await this.setActiveServerDatabase({serverUrl, displayName: databaseName});
      }

      return connection;
  }

  /**
   * createDatabaseConnection: Creates database connection and registers the new connection into the default database.  However,
   * if a database connection could not be created, it will return undefined.
   * @param {MMDatabaseConnection} databaseConnection
   * @param {boolean} shouldAddToDefaultDatabase
   *
   * @returns {Promise<DatabaseInstance>}
   */
  createDatabaseConnection = async ({configs, shouldAddToDefaultDatabase = true}: DatabaseConnectionArgs): Promise<DatabaseInstance> => {
      const {actionsEnabled = true, dbName = DEFAULT_DATABASE, dbType = DatabaseType.DEFAULT, serverUrl = undefined} = configs;

      try {
          const databaseName = dbType === DatabaseType.DEFAULT ? DEFAULT_DATABASE : dbName;

          const databaseFilePath = this.getDatabaseDirectory(databaseName);
          const migrations = dbType === DatabaseType.DEFAULT ? DefaultMigration : ServerMigration;
          const modelClasses = dbType === DatabaseType.DEFAULT ? this.defaultModels : this.serverModels;
          const schema = dbType === DatabaseType.DEFAULT ? defaultSchema : serverSchema;

          const adapter = new SQLiteAdapter({
              dbName: databaseFilePath,
              migrationEvents: this.buildMigrationCallbacks(databaseName),
              migrations,
              schema,
          });

          // Registers the new server connection into the DEFAULT database
          if (serverUrl && shouldAddToDefaultDatabase) {
              await this.addServerToDefaultDatabase({
                  databaseFilePath,
                  displayName: dbName,
                  serverUrl,
              });
          }

          return new Database({adapter, actionsEnabled, modelClasses});
      } catch (e) {
          // TODO : report to sentry? Show something on the UI ?
      }

      return undefined;
  };

  /**
   * setActiveServerDatabase: Set the new active server database.  The serverUrl is used to ensure that we do not duplicate entries in the default database.
   * This method should be called when switching to another server.
   * @param {string} displayName
   * @param {string} serverUrl
   * @returns {Promise<void>}
   */
  setActiveServerDatabase = async ({displayName, serverUrl}: ActiveServerDatabaseArgs) => {
      const isServerPresent = await this.isServerPresent(serverUrl);

      //fixme: use a 'flag' under the GLOBAL entity called 'recentlyUsedServer'. This flag will contain an array of recently connected servers and the server at index 0 will be the last server visited/viewed
      this.activeDatabase = await this.createDatabaseConnection({
          configs: {
              actionsEnabled: true,
              dbName: displayName,
              dbType: DatabaseType.SERVER,
              serverUrl,
          },
          shouldAddToDefaultDatabase: Boolean(!isServerPresent),
      });
  };

  /**
   * isServerPresent : Confirms if the current serverUrl does not already exist in the database
   * @param {String} serverUrl
   * @returns {Promise<boolean>}
   */
  private isServerPresent = async (serverUrl: string) => {
      const allServers = await this.getAllServers([serverUrl]);

      // const existingServer = allServers?.filter((server) => {
      //     return server.url === serverUrl;
      // });

      return allServers?.length > 0;
  };

  /**
   * getActiveServerDatabase: The DatabaseManager should be the only one setting the active database.  Hence, we have made the activeDatabase property private.
   * Use this getter method to retrieve the active database if it has been set in your code.
   * @returns {DatabaseInstance}
   */
  getActiveServerDatabase = (): DatabaseInstance => {
      //fixme: read index 0 from 'flag' from GLOBAL entity  and return it.
      return this.activeDatabase;
  };

  /**
   * getDefaultDatabase : Returns the default database.
   * @returns {Database} default database
   */
  getDefaultDatabase = async (): Promise<DatabaseInstance> => {
      if (!this.defaultDatabase) {
          await this.setDefaultDatabase();
      }
      return this.defaultDatabase;
  };

  /**
   * retrieveDatabaseInstances: Using an array of server URLs, this method creates a database connection for each URL
   * and return them to the caller.
   *
   * @param {string[]} serverUrls
   * @returns {Promise<RetrievedDatabase[] | null>}
   */
  retrieveDatabaseInstances = async (serverUrls: string[]): Promise<DatabaseInstances[] | null> => {
      // Retrieve all server records from the default db
      const allServers = await this.getAllServers(serverUrls);

      // Creates server database instances
      if (allServers.length) {
          const databasePromises = await allServers.map(async (server: IServers) => {
              const {displayName, url} = server;

              // Since we are retrieving existing URL ( and so database connections ) from the 'DEFAULT' database, shouldAddToDefaultDatabase is set to false
              const dbInstance = await this.createDatabaseConnection({
                  configs: {
                      actionsEnabled: true,
                      dbName: displayName,
                      dbType: DatabaseType.SERVER,
                      serverUrl: url,
                  },
                  shouldAddToDefaultDatabase: false,
              });

              return {dbInstance, displayName, url};
          });
          const databaseInstances = await Promise.all(databasePromises);
          return databaseInstances;
      }
      return null;
  };

  /**
   * deleteDatabase: Removes the *.db file from the App-Group directory for iOS or the files directory on Android.
   * Also, it removes its entry in the 'servers' table from the DEFAULT database
   * @param  {string} serverUrl
   * @returns {Promise<boolean>}
   */
  deleteDatabase = async (serverUrl: string): Promise<boolean> => {
      try {
          const defaultDB = await this.getDefaultDatabase();
          let server: IServers;

          if (defaultDB) {
              // NOTE: We are deleting this 'database' entry in the SERVERS entity on the DEFAULT database; for that we retrieve its record first.
              const serversRecords = (await defaultDB.collections.get(SERVERS).query(Q.where('url', serverUrl)).fetch()) as IServers[];
              server = serversRecords?.[0] ?? undefined;

              if (server) {
                  const databaseName = server.displayName;

                  // Perform a delete operation for this server record on the 'servers' table in default database
                  await defaultDB.action(async () => {
                      await server.destroyPermanently();
                  });

                  if (Platform.OS === 'ios') {
                      // On iOS, we'll delete the *.db file under the shared app-group/databases folder
                      deleteIOSDatabase({databaseName});
                      return true;
                  }

                  // On Android, we'll delete both the *.db file and the *.db-journal file
                  const androidFilesDir = `${this.androidFilesDirectory}databases/`;
                  const databaseFile = `${androidFilesDir}${databaseName}.db`;
                  const databaseJournal = `${androidFilesDir}${databaseName}.db-journal`;

                  await FileSystem.deleteAsync(databaseFile);
                  await FileSystem.deleteAsync(databaseJournal);

                  return true;
              }
              return false;
          }
          return false;
      } catch (e) {
          return false;
      }
  };

  /**
   * factoryReset: Removes the databases directory and all its contents on the respective platform
   * @param {boolean} shouldRemoveDirectory
   * @returns {Promise<boolean>}
   */
  factoryReset = async (shouldRemoveDirectory: boolean): Promise<boolean> => {
      try {
          //On iOS, we'll delete the databases folder under the shared AppGroup folder
          if (Platform.OS === 'ios') {
              deleteIOSDatabase({shouldRemoveDirectory});
              return true;
          }

          // On Android, we'll remove the databases folder under the Document Directory
          const androidFilesDir = `${FileSystem.documentDirectory}databases/`;
          await FileSystem.deleteAsync(androidFilesDir);
          return true;
      } catch (e) {
          return false;
      }
  };

  /**
   * getAllServers : Retrieves all the servers registered in the default database
   * @returns {Promise<Servers[]>}
   */
  private getAllServers = async (serverUrls: string[]) => {
      // Retrieve all server records from the default db
      const defaultDatabase = await this.getDefaultDatabase();

      if (defaultDatabase) {
          const allServers = (await defaultDatabase.collections.get(SERVERS).query(Q.where('url', Q.oneOf(serverUrls))).fetch() as IServers[]);
          return allServers;
      }

      return [];
  };

  /**
   * setDefaultDatabase : Sets the default database.
   * @returns {Promise<DatabaseInstance>}
   */
  private setDefaultDatabase = async (): Promise<DatabaseInstance> => {
      this.defaultDatabase = await this.createDatabaseConnection({
          configs: {dbName: DEFAULT_DATABASE},
          shouldAddToDefaultDatabase: false,
      });

      return this.defaultDatabase;
  };

  /**
   * addServerToDefaultDatabase: Adds a record into the 'default' database - into the 'servers' table - for this new server connection
   * @param {string} databaseFilePath
   * @param {string} displayName
   * @param {string} serverUrl
   * @returns {Promise<void>}
   */
  private addServerToDefaultDatabase = async ({databaseFilePath, displayName, serverUrl}: DefaultNewServerArgs) => {
      try {
          const defaultDatabase = await this.getDefaultDatabase();
          const isServerPresent = await this.isServerPresent(serverUrl);

          if (defaultDatabase && !isServerPresent) {
              await defaultDatabase.action(async () => {
                  const serversCollection = defaultDatabase.collections.get(SERVERS);
                  await serversCollection.create((server: IServers) => {
                      server.dbPath = databaseFilePath;
                      server.displayName = displayName;
                      server.mentionCount = 0;
                      server.unreadCount = 0;
                      server.url = serverUrl;
                  });
              });
          }
      } catch (e) {
          // TODO : report to sentry? Show something on the UI ?
      }
  };

  /**
   * buildMigrationCallbacks: Creates a set of callbacks that can be used to monitor the migration process.
   * For example, we can display a processing spinner while we have a migration going on. Moreover, we can also
   * hook into those callbacks to assess how many of our servers successfully completed their migration.
   * @param {string} dbName
   * @returns {MigrationEvents}
   */
  private buildMigrationCallbacks = (dbName: string) => {
      const migrationEvents: MigrationEvents = {
          onSuccess: () => {
              return DeviceEventEmitter.emit(MIGRATION_EVENTS.MIGRATION_SUCCESS, {
                  dbName,
              });
          },
          onStarted: () => {
              return DeviceEventEmitter.emit(MIGRATION_EVENTS.MIGRATION_STARTED, {
                  dbName,
              });
          },
          onFailure: (error) => {
              return DeviceEventEmitter.emit(MIGRATION_EVENTS.MIGRATION_ERROR, {
                  dbName,
                  error,
              });
          },
      };

      return migrationEvents;
  };

  /**
   * getDatabaseDirectory: Using the database name, this method will return the database directory for each platform.
   * On iOS, it will point towards the AppGroup shared directory while on Android, it will point towards the Files Directory.
   * Please note that in each case, the *.db files will be created/grouped under a 'databases' sub-folder.
   * iOS Simulator : appGroup => /Users/{username}/Library/Developer/CoreSimulator/Devices/DA6F1C73/data/Containers/Shared/AppGroup/ACA65327/databases"}
   * Android Device: file:///data/user/0/com.mattermost.rnbeta/files/databases
   *
   * @param {string} dbName
   * @returns {string}
   */
  private getDatabaseDirectory = (dbName: string): string => {
      return Platform.OS === 'ios' ? `${this.iOSAppGroupDatabase}/${dbName}.db` : `${FileSystem.documentDirectory}${dbName}.db`;
  };
}

if (!__DEV__) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    logger.silence();
}

export default new DatabaseManager();
