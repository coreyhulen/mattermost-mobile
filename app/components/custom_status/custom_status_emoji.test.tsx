// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';

import {SYSTEM_IDENTIFIERS} from '@constants/database';
import DatabaseManager from '@database/manager';
import ServerDataOperator from '@database/operator/server_data_operator';
import NetworkManager from '@init/network_manager';
import {MOCKED_DATA} from '@test/mock_database_data';
import {renderWithDatabase} from '@test/intl-test-helper';

import CustomStatusEmoji from './custom_status_emoji';

import type {Database} from '@nozbe/watermelondb';

describe('@components/custom_status/custom_status_emoji', () => {
    let database: Database;
    let operator: ServerDataOperator;

    beforeAll(async () => {
        const serverUrl = 'baseHandler.test.com';
        await DatabaseManager.init([serverUrl]);
        database = DatabaseManager.serverDatabases[serverUrl].database;
        operator = DatabaseManager.serverDatabases[serverUrl].operator;

        await NetworkManager.init([{serverUrl, userId: '1234567', token: '432424788^^83434'}]);

        const users = operator.handleUsers({
            users: [
                {
                    id: 'p9g6rzz3kffhxqxhm1zckjpwda',
                    create_at: 1599457495881,
                    update_at: 1607683720173,
                    delete_at: 0,
                    username: 'a.l',
                    auth_service: 'saml',
                    email: 'a.l@mattermost.com',
                    email_verified: true,
                    is_bot: false,
                    nickname: '',
                    first_name: 'A',
                    last_name: 'L',
                    position: 'Mobile Engineer',
                    roles: 'system_user',
                    props: {},
                    notify_props: {
                        mark_unread: 'all',
                        desktop: 'all',
                        desktop_sound: 'true',
                        email: 'true',
                        first_name: 'true',
                        mention_keys: '',
                        push: 'mention',
                        channel: 'true',
                        auto_responder_active: 'false',
                        auto_responder_message:
                            'Hello, I am out of office and unable to respond to messages.',
                        comments: 'never',
                        desktop_notification_sound: 'Hello',
                        push_status: 'online',
                    },
                    last_picture_update: 1604686302260,
                    locale: 'en',
                    timezone: {
                        automaticTimezone: 'Indian/Mauritius',
                        manualTimezone: '',
                        useAutomaticTimezone: '',
                    },
                },
                {
                    id: '1234567',
                    create_at: 1599457495881,
                    update_at: 1607683720173,
                    delete_at: 0,
                    username: 'a.l',
                    auth_service: 'saml',
                    email: 'a.l@mattermost.com',
                    email_verified: true,
                    is_bot: false,
                    nickname: '',
                    first_name: 'A',
                    last_name: 'L',
                    position: 'Mobile Engineer',
                    roles: 'system_user',
                    props: {
                        customStatus: {
                            emoji: 'calendar',
                            text: 'In a meeting',
                        },
                    },
                    notify_props: {
                        desktop: 'all',
                        mark_unread: 'all',
                        desktop_sound: 'true',
                        email: 'true',
                        first_name: 'true',
                        mention_keys: '',
                        push: 'mention',
                        channel: 'true',
                        auto_responder_active: 'false',
                        auto_responder_message:
                            'Hello, I am out of office and unable to respond to messages.',
                        comments: 'never',
                        desktop_notification_sound: 'Hello',
                        push_status: 'online',
                    },
                    last_picture_update: 1604686302260,
                    locale: 'en',
                    timezone: {
                        automaticTimezone: 'Indian/Mauritius',
                        manualTimezone: '',
                        useAutomaticTimezone: '',
                    },
                }],
            prepareRecordsOnly: true,
        });

        const systems = operator.handleSystem({
            systems: [
                {id: 'currentUserId', value: 'p9g6rzz3kffhxqxhm1zckjpwda'},
                {id: SYSTEM_IDENTIFIERS.CONFIG, value: MOCKED_DATA.CONFIG},
            ],
            prepareRecordsOnly: true,
        });

        const emojis = operator.handleCustomEmojis({
            emojis: [
                {
                    id: 'i',
                    create_at: 1580913641769,
                    update_at: 1580913641769,
                    delete_at: 0,
                    creator_id: '4cprpki7ri81mbx8efixcsb8jo',
                    name: 'calendar',
                },
            ],
            prepareRecordsOnly: true,
        });

        const models = await Promise.all([emojis, systems, users]);

        const flattenedModels = models.flat();
        if (flattenedModels?.length > 0) {
            await operator.batchRecords(flattenedModels);
        }
    });

    it('should match snapshot', () => {
        const wrapper = renderWithDatabase(<CustomStatusEmoji/>, database);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should return undefined for currentUserId as no custom status is set', () => {
        const wrapper = renderWithDatabase(<CustomStatusEmoji/>, database);
        expect(wrapper.toJSON()).toBeNull();
    });

    it('should return a value for userId 1234567 as a custom status is set', () => {
        const wrapper = renderWithDatabase(<CustomStatusEmoji userId={'1234567'}/>, database);
        expect(wrapper.toJSON()).toBeDefined();
    });
});
