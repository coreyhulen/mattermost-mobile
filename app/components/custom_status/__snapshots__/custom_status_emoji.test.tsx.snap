// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/custom_status/custom_status_emoji should match snapshot 1`] = `null`;
